### PR TITLE
Change font-size-picker markup to use select

### LINKED ISSUE #15319: Custom interactions in control styled as a native control
Custom interactions in control styled as a native control

-   **Severity:** Medium
-   **Affected Populations:**
    -   Low-Vision
    -   Motor Impaired
    -   Cognitively Impaired
-   **Platform(s):**
    -   All / Universal
-   **Components affected:**
    -   Block Panel

#### Issue description

In the Block panel, where the font size can be selected, users see what
appears to be a select element. However when keyboard users reach this
component, they cannot use standard keystrokes to change the value, such
as Up-Arrow or Down-Arrow, or Alt+Arrow to open the submenu. Instead,
users need to press Enter to view the submenu.

Keyboard users will expect this control to work the same as a native
select element. Speech recognition users often speak out keystrokes to
perform standard actions on native components, and these would not work
on custom components like these.

##### Issue Code

```
    <button type="button" aria-expanded="false" aria-label="Font size: Normal" class="components-button components-font-size-picker__selector is-button is-default is-large">Normal</button>

```

#### Remediation Guidance

Use a native `<select>` element in place of a custom button+submenu
setup.

##### Recommended Code

```
    <fieldset class="components-base-control__field">


        <legend class="components-base-control__label">Font Size</legend>


        <label for="font-size-picker">Choose basic size</label>


        <select id-"font-size-picker" class="attachment-filters">


            <option value="normal" >Normal</option>


            ...


        </select>


        <label for="custom-font-size">Custom font size</label>


        <input id="custom-font-size" class="components-range-control__number" type="number" value="">


        <button type="button" disabled="" class="components-button components-color-palette__clear is-button is-default is-small">Reset</button>


    </fieldset>
```

#### Relevant standards

-   [2.1.1 Keyboard
    (Level A)](https://www.w3.org/TR/WCAG20/#keyboard-operation-keyboard-operable)

**Note:** This issue may be a duplicate with other existing accessibility-related bugs in this project. This issue comes from the Gutenberg accessibility audit, performed by [Tenon](https://www.tenon.io) and funded by [WP Campus](https://wpcampus.org/). This issue is GUT-68 in Tenon's report

### PR DESCRIPTION
**Pull request by tellthemachines**
## Description

Fixes #15319

Changed font-size-picker markup to use fieldset with legend as wrapper and replaced visual dropdown with select. 

## How has this been tested?

Tested locally on Chrome and on Safari with VoiceOver. Confirmed interaction with controls works as expected. Also tested on Windows with Firefox + NVDA; Noted NVDA doesn't read out the "Font size" legend before each field but only when entering the fieldset; VoiceOver reads it out before each field.

Wondering if it might be worth adding a snapshot and/or an e2e test for this. 

## Screenshots 

Screen Shot 2019-06-13 at 5 53 07 pm

Screen Shot 2019-06-13 at 5 56 36 pm

## Types of changes

Bug fix (non-breaking change which fixes an issue)

* Changes to markup, css, state and value handling within `font-size-picker` component.
* Changes to `base-control` and ` select-control` components in order to allow control label to be visually hidden.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by afercia** [comment]
@tellthemachines thanks for working on this!

From an accessibility perspective, everything looks good to me 👍 
- a native `<select>` element is very welcome
- the fieldset legend makes sense when announced in combination with the input labels

As explained in related issues, it's important to consider the _combination_ of the legend text and the various input labels. In this specific case, removing "font size" from the labels avoids redundant (and a bit pointless) information to be announced to users, as the group is already named "font size".

> NVDA doesn't read out the "Font size" legend before each field but only when entering the fieldset; VoiceOver reads it out before each field.

Yep, implementations differ and actually the most correct behavior is the Firefox + NVDA one 🙂 The group name should be announced only when entering and exiting the group. 

Note: the fieldset legend styling should be quickly tested with Internet Explorer 11, as I seem to recall it was a bit difficult to style in that browser.

The build fails, not sure why.

**Comment by tellthemachines** [comment]
Thanks @afercia for the feedback! I changed the selectors on the e2e tests and now they are passing.

File: packages/components/src/base-control/index.js [hunk 1]
```diff
@@ -3,11 +3,11 @@
  */
 import classnames from 'classnames';
 
-function BaseControl( { id, label, help, className, children } ) {
+function BaseControl( { id, label, hideLabel, help, className, children } ) {
```
Comment (gziolo) [inline packages/components/src/base-control/index.js]: `hideLabel` - this new prop has to be documented. I'm a bit afraid that the name proposed when read outside of its implementation might be confusing. Based on: https://make.wordpress.org/accessibility/handbook/markup/the-css-class-screen-reader-text/:

> The screen-reader-text class is used:
> * to hide text from vision that is only useful for screen readers users.
> * for skip links for keyboard navigation before the link gets focus.

It should be something like `hideLabelFromVision`.

File: packages/components/src/font-size-picker/index.js [hunk 2]
```diff
@@ -26,76 +31,61 @@ function FontSizePicker( {
...
-							</NavigableMenu>
-						) }
+					<SelectControl
+						className={ 'components-font-size-picker__select' }
+						label={ 'Choose preset' }
+						hideLabel={ true }
+						value={ currentSelectValue }
+						onChange={ onSelectChangeValue }
+						options={ [
+							{ value: fontSizes[ 0 ].slug, label: __( 'Small' ) },
```
Comment (gziolo) [inline packages/components/src/font-size-picker/index.js]: Font sizes can be customized for each site: 
https://developer.wordpress.org/block-editor/developers/themes/theme-support/#block-font-sizes

The names used as labels and the number of items may differ.

**Review by gziolo — DISMISSED** [review]

File: packages/components/src/font-size-picker/index.js [hunk 3]
```diff
@@ -26,76 +31,61 @@ function FontSizePicker( {
 	value,
 	withSlider = false,
 } ) {
+	// eslint-disable-next-line @wordpress/no-unused-vars-before-return
+	const [ currentSelectValue, setCurrentSelectValue ] = useState( getInitialSelectValue( fontSizes, value ) );
```
Comment (gziolo) [inline packages/components/src/font-size-picker/index.js]: It seems like this changes the behavior of the control. It no longer will get its value updated when someone manually enters the value in the input. The idea is that this combo box should also update when someone picks the font size which matches one of the presets.

File: packages/components/src/font-size-picker/index.js (hunk 2 above)
Comment (tellthemachines) [reply packages/components/src/font-size-picker/index.js]: I tried to generate `options` dynamically from the `fontSizes` array but it seems the i18n `__` function can't be called with dynamic content. Is there any workaround for this?

File: packages/components/src/font-size-picker/index.js (hunk 3 above)
Comment (tellthemachines) [reply packages/components/src/font-size-picker/index.js]: Oh, that functionality is working for me locally. It doesn't update if you empty the input box, but if you add any value to it the select updates to the correct option. Could you give me more details about what you tried so I can try to reproduce it? Thanks!

File: packages/components/src/font-size-picker/index.js (hunk 2 above)
Comment (gziolo) [reply packages/components/src/font-size-picker/index.js]: When those options are set on the server, they should be already translated. There is no need to wrap them with `__` on the client.

File: packages/components/src/font-size-picker/index.js (hunk 3 above)
Comment (gziolo) [reply packages/components/src/font-size-picker/index.js]: In `master` branch when you reset value using the reset button or by emptying the input box it reflects with the font size update to the default name – Normal – in case of Paragraph block. It doesn't work this way in this branch. There is one source of truth in the previous implementation – `value` – which can be updated with `onChange` helper. In this branch, there is a local state introduced which shadows the same value but in my testing, it doesn't fully replicate the old behavior. This component updated every time `value` changes, so it might be enough to keep it working similar to what existed before.

File: packages/block-editor/src/components/font-sizes/with-font-sizes.js [hunk 4]
```diff
@@ -64,7 +64,7 @@ export default ( ...fontSizeNames ) => {
 
 					createSetFontSize( fontSizeAttributeName, customFontSizeAttributeName ) {
 						return ( fontSizeValue ) => {
-							const fontSizeObject = find( this.props.fontSizes, { size: fontSizeValue } );
+							const fontSizeObject = find( this.props.fontSizes, { size: Number( fontSizeValue ) } );
```
Comment (jorgefilipecosta) [inline packages/block-editor/src/components/font-sizes/with-font-sizes.js]: Hi, what was the reason for this change? fontSizeValue should already be a number right?

File: packages/components/src/font-size-picker/index.js (hunk 2 above)
Comment (tellthemachines) [reply packages/components/src/font-size-picker/index.js]: Got it! Thanks.

File: packages/block-editor/src/components/font-sizes/with-font-sizes.js (hunk 4 above)
Comment (tellthemachines) [reply packages/block-editor/src/components/font-sizes/with-font-sizes.js]: The value retrieved from the select is [`event.target.value`](https://github.com/WordPress/gutenberg/blob/090586e6273dbd637b167d87b37b977e32a5a23f/packages/components/src/select-control/index.js#L34) which is a string.

File: packages/components/src/font-size-picker/index.js (hunk 3 above)
Comment (tellthemachines) [reply packages/components/src/font-size-picker/index.js]: It should be following previous behaviour exactly now!

File: packages/components/src/base-control/index.js [hunk 5]
```diff
@@ -3,11 +3,11 @@
  */
 import classnames from 'classnames';
 
-function BaseControl( { id, label, help, className, children } ) {
+function BaseControl( { id, label, hideLabelFromVision, help, className, children } ) {
 	return (
 		<div className={ classnames( 'components-base-control', className ) }>
 			<div className="components-base-control__field">
-				{ label && id && <label className="components-base-control__label" htmlFor={ id }>{ label }</label> }
+				{ label && id && <label className={ classnames( 'components-base-control__label', hideLabelFromVision && 'screen-reader-text' ) } htmlFor={ id }>{ label }</label> }
```
Comment (talldan) [inline packages/components/src/base-control/index.js]: This line is getting quite busy, it would be great to break it up onto multiple lines. The call to classnames could also be written like this, which is a pattern used more commonly across the codebase:
```jsx
{ classnames( 'components-base-control__label', {
    'screen-reader-text': hideLabelFromVision,
} ) }
```

File: packages/components/src/font-size-picker/index.js [hunk 6]
```diff
@@ -1,22 +1,34 @@
+
 /**
- * External dependencies
+ * WordPress dependencies
```
Comment (talldan) [inline packages/components/src/font-size-picker/index.js]: Looks like there are two WordPress dependency comments, would be great to merge the imports into one section.

File: packages/components/src/font-size-picker/index.js [hunk 7]
```diff
@@ -26,84 +38,64 @@ function FontSizePicker( {
...
+	// eslint-disable-next-line @wordpress/no-unused-vars-before-return
+	const [ currentSelectValue, setCurrentSelectValue ] = useState( getInitialSelectValue( fontSizes, value ) );
+
 	if ( disableCustomFontSizes && ! fontSizes.length ) {
 		return null;
 	}
+
 	const onChangeValue = ( event ) => {
 		const newValue = event.target.value;
+		setCurrentSelectValue( getInitialSelectValue( fontSizes, Number( newValue ) ) );
```
Comment (talldan) [inline packages/components/src/font-size-picker/index.js]: With `getInitialSelectValue` being used here in `onChangeValue` as well, its name no longer makes as much sense—it's no longer being used for just the initial value. Maybe something like `getSelectValueFromFontSize` could work better.

**Review by talldan — APPROVED** [review]
This looks good to me. I had a few minor comments, but nothing major. Would be good to tidy up the comment above that import, but the other two comments are more subjective.

File: packages/components/src/font-size-picker/index.js (hunk 6 above)
Comment (tellthemachines) [reply packages/components/src/font-size-picker/index.js]: 🤦‍♂ how ever did that happen?

File: packages/components/src/font-size-picker/index.js (hunk 7 above)
Comment (tellthemachines) [reply packages/components/src/font-size-picker/index.js]: Done!

**Comment by tellthemachines** [comment]
@mapk @kjellr could I get your tick of approval (or design feedback) on this please?

**Comment by karmatosed** [comment]
If the decision is to use a select box then this works. I do wonder what the boundaries are on us styling that though, that's perhaps something to explore going forward. @afercia would you be able to advise on where those boundaries are?

**Comment by kjellr** [comment]
> If the decision is to use a select box then this works. I do wonder what the boundaries are on us styling that though, that's perhaps something to explore going forward. @afercia would you be able to advise on where those boundaries are?

I actually looked into this briefly a couple weeks ago. This seems to be mosly up to date: 

https://css-tricks.com/dropdown-default-styling/

In general: Webkit browsers, and Firefox on Mac do not allow much styling. IE, Firefox, and Opera allow you to do a bit of minimal styling on Windows and Linux. (But I imagine that'll change a bit once the next version of IE adopts webkit.) Also, iOS does not allow you to style select fields. I do not have an Android phone handy at the moment, but I do not believe they support styling them either. 

In short: if we move forward with a select box here, we shouldn't count on styling it.

**Comment by afercia** [comment]
At Yoast we do use a component that's just a _native_ styled select element. Basically, only the select "box" can be styled consistently across desktop browsers (including IE 11). The dropdown with options can't. Not sure about mobile browsers. Screenshots (Chrome on macOS - dark theme):

Screenshot 2019-07-08 at 15 24 20

Screenshot 2019-07-08 at 15 24 32

**Comment by enriquesanchez** [comment]
While I have never used nor tested it, this select component from Deque's patter library seems to be fully accessible while also allowing for custom styling: https://pattern-library.dequelabs.com/components/selects.

Could something like that work for us?

**Comment by afercia** [comment]
That looks like an excellent implementation. Still not native though. Off the top of my head, I can spot a couple issues at a first glance:
- support for `aria-haspopup="listbox"` is still poor
- to my knowledge, speech recognition software (Dragon) have troubles with `aria-labelledby` thus voice commands won't work

Maybe @dboudreau can help and give us some updated feedback or point us to someone at deque (when you have a chance, that would be greatly appreciated!). Background: evaluating to use the deque select component in WordPress.

**Comment by noisysocks** [comment]
A custom select component is worth exploring but out of scope for this issue. The decision in https://github.com/WordPress/gutenberg/issues/15319#issuecomment-500623267 was to use a native `<select>`, so let's get this in as is and iterate! 😀
